### ISSUE TITLE
Integers in HoTT

### ISSUE THREAD

**Issue by Alizter** [HoTT/HoTT#1124]
I propose we start collecting different definitions of the integers in a folder in Spaces. Currently we have two main definitions floating around:
 * The integers as N + 1 + N defined in Spaces.Int.
 * The integers as N * N / ~ defined in natpair_integers.v in Classes/implementations

I think it will be beneficial for us to put these in one place, with some auxiliary files showing that they are all somehow equivalent.

Now one of the interesting things about HoTT is that we can come up with some more definitions of the integers. There are a few directly as HITs and of course another as the loop space of the circle.

Currently if somebody says: "Hey, I need to use the integers" they have to choose between Spaces.Int and natpair_integers. The definition in Spaces.Int is a bit tricky to work with, we don't even have a definition of addition or multiplication yet. And the natpair version is hidden away in hottclasses for everybody to miss.

Personally, I have a hard time reading the code in the natpair version too, since I am not used to that style. Not to mention, when using natpair, you have to import the NatPair module. OK, fair enough, we import modules for modalities. But then when you want to talk about the type Z, you have to specify which naturals too.

This seems like a very convoluted way to talk about the integers, and is definitely not what I would call easy to use. I instead propose we define the integers like we do in the HoTT book as the quotient of the naturals. I think natpair is doing the same thing but in a very general way which makes it difficult to use and understand.

I understand some of the motivation behind this generality however, since when you can provide whichever implementation of the integers or naturals you can plug in the efficient versions for computations.

I think we need to have a discussion over how we want to implement the integers.

**Comment by mikeshulman** [comment]
While it's a cute idea to define "a notion of integer" as hottclasses does that can be "implemented" in many ways, I feel like for practical use we ought to just pick one definition of "the integers" and stick with it.  We already have enough transporting across equivalences, I don't want the problem of putting together two different developments to be made more difficult by the fact that they use different notions of integer.  The possibility of the other definitions can be made visible by, for each definition, proving that our chosen notion of integer satisfies that definition (e.g. has the appropriate induction principle), and moreover that any type satisfying that definition is equivalent to our chosen one.

This is what we did with equivalences, for instance: we chose half-adjoint equivalences in defining `IsEquiv` and then in `EquivalenceVarieties` we proved that all the other notions of equivalence are equivalent, but in practice everywhere in the library that we say `IsEquiv` we mean half-adjoint ones, and if someone constructs an equivalence using some other definition they immediately make it into a half-adjoint one using the results in `EquivalenceVarieties`.

If we care about efficiency for computations (which we might one day, even if we don't now), why not just choose an efficient implementation to start with as "the integers"?  I do think we should stick with the unary naturals as "the naturals", but for the integers why not go with binary ones as e.g. in 
the coq stdlib?  I think there are advantages to choosing a definition that doesn't involve any HITs.

**Comment by Alizter** [comment]
The "binary integers" in the coq stdlib are implemented as follows:

 1. There is a data type called `positive` which represents positive binary integers.
 2. Naturals N are defined to be `positive + 1`.
 3. Integers are defined to be `N + 1 + N`.

So this is pretty much how we have already defined the integers with one key difference. The binary naturals are used instead of the unary ones.

Confusingly in Spaces.Int we have a datatype called `Pos` which is just the unary naturals. If we really want to mimic the coq library, we should replace this with a version of the binary naturals. This shouldn't be too difficult.

I would rather refrain from relying on the coq stdlib. Since we only need basic arithmetic from that I would be in favour of duplicating that in some appropriate way.

And while I am talking about that point, I think the same should be done for the naturals too as it currently relies on coq's nat. I don't really see why we want to rely on coq's unary naturals if we can do it ourselves.

**Comment by mikeshulman** [comment]
I don't think there's any problem with using things from the stdlib if they make sense for us.  Why duplicate effort?

**Comment by Alizter** [comment]
@mikeshulman Here are a few reasons I could think of:
 1. There is not that much to duplicate anyway.
 2. We don't become bound to any changes to the main library (however unlikely).
 3. Having everything in one library makes it easier for people to read. So the organisation of the library becomes more logical.
 4. We can adhere nat closer to our naming conventions.

The third reason is especially important for newcomers to the library who don't know anything about coq, or where to look for standard library things.

Apart from not having to write all this code, I don't see any other advantage to not duplicating. Also I would be happy to take on this task.

**Comment by mikeshulman** [comment]
Why don't we try just copy-pasting the integers code from the standard library, and maybe change the names here and there?  Then we can save the effort of writing it ourselves but keep control of our organization and naming conventions.

**Comment by Alizter** [comment]
@mikeshulman That's essentially what I had in mind with "writing it ourselves". :-)

**Comment by mikeshulman** [comment]
Ah, okay.  That's fine with me, I guess.  Unless anyone else has an opinion?

**Comment by Alizter** [comment]
After some investigation I don't think changing nat is really an option. Coq considers this a fundamental data type which is loaded every time. So I think we are stuck with that.

For the integers however we can definitely do as I suggested above, and have our "HoTT specific" version.

I suppose one way to remedy the nat case is to better document what the coq library proves for nats so people don't have to go searching for lemmas outside of the HoTT library.

**Comment by spitters** [comment]
For maintainability, I think we'd like to stay as close to the stdlib as
possible.
There's been a lot of improvements, as more tactics work now with equality
in type.

**Comment by Alizter** [comment]
OK I am now convinced that even just copying the binary integers is a big pain. There are all sorts of tricks that the library uses in order to display notations etc.

I am happy with using Coq.Numbers.BigNums.Z in Spaces.Int instead of defining it ourselves. I can't seem to access this part of the coq library however.

**Comment by mikeshulman** [comment]
Most of the Coq stdlib was explicitly ripped out of the HoTT library; if you look at coq/theories most of the directories are empty.  We'll have to put back manually whatever we want to have.

I would prefer to name the integers `Int` rather than `Z`, though, if we can manage that...

**Comment by Alizter** [comment]
Actually if we can access Z from the coq library, we can continue to mirror its definition ourselves. I think I have a way to get the numeral notations working for our Int, but it will require access to Z.

**Comment by Alizter** [comment]
I am actually quite suprised how we have access to the coq libraries. I thought we were just looking at the ones from the coq version the library is configured for.

I am not wholly convinced that keeping a small shadow of the coq library along with us is any better than just copying some files, changing them where needed and adding them to the coq library?

For example, we can quite literally rename nat to Nat, since this isn't the "real" coq nat. But then this begs the question why we keep it there at all, why not move all of this into Spaces.Nat?

**Comment by spitters** [comment]
We (used to) need a copy of the coq library because some lemmas were hard
coded in the coq tactics.
This has improved, but I am not sure by how much.

Generally, for maintainability, we'd like to be as close to standard coq as
possible.

**Comment by Alizter** [comment]
@spitters How does being close to standard coq aid with maintainability?

**Comment by spitters** [comment]
It would be nice to have tactics like omega, lia, ... to work.
HoTT has little new to add about discrete types, so, maybe we can join
forces with the coq-dev maintaining that.

**Comment by spitters** [comment]
Ideally stdlib2 would be hott compatible...
https://github.com/coq/stdlib2/

**Comment by Alizter** [comment]
@spitters Thank you for your comments. I think we have some more to discuss about how close we want to be with coq in the future. I will open another issue to discuss this in the future.

For now however, I will see about adding the coq version of binary integers into our coq shadow. Then on top of that we can develop our own integers for the hott library.

**Comment by Alizter** [comment]
OK a quick update on how I am doing. I couldn't get the coq version of the integers working, since it requires a lot more of the coq library and various tactics too.

I have instead settled for copying the definition of positive binary number and integer defined as before but using the positive binary numbers.

Notably I have managed to define numeral notations for positive binary numbers and integers. I will now work on proving associativity of various operations, which seems to be pretty difficult(/tedious) and done on a case by case basis.

One of the difficulties with the binary positive numbers is that they have 3 constructors (one, a left 1 bit, a left 0 bit). This causes many cases to quickly multiple. I should be able to prove the induction principle with the base and successor but I haven't gotten round to doing that yet.

Slow but steady progress.

**Comment by mikeshulman** [comment]
Thanks for working on this.  Sometimes when working with datatypes that have a lot of constructors, some clever use of tactics and tacticals can shorten the proofs dramatically, but I don't know whether that's the case here.

**Comment by Alizter** [comment]
@mikeshulman I agree that the proofs can probably be shortened, since it is longer than the coq library version. Hopefully somebody in the future will have a slicker way of doing things. For now, my version will get us by.

**Comment by JasonGross** [comment]
Note that the Coq standard library is incompatible with `-indices-matter`.  I've never been able to get an explicit contradiction from assuming that every equivalence type `A <~> B` is itself equivalent to some type in `Set`, but it wouldn't surprise me if you could get a version of Hurkens paradox by assuming `(Set <~> Set) : Set` (I don't immediately see how, though).  Given that, we need to include our own copy of the standard library built with `-indices-matter`.

**Comment by Alizter** [comment]
@JasonGross I had some vague understanding about incompatibilities with the main coq library. Thanks for mentioning those examples.

Do you know how to get negative numerals working with `Numeral Notation`? I have managed to define functions from our `Int` to `Decimal.int` and vice versa, being careful about what is contained in the bodies of the definitions. It seems to be working for positive integers, however I cannot get zero to print or any negative numbers for that matter.

My solution is to just define a notation for zero and the negatives, which seems to make it print fine. However I cannot get coq to parse the following: `-2`, but it will parse something like `- 2`. When printing `- 2` it writes `-2`. 

Am I doing something wrong here?

**Comment by spitters** [comment]
@andreaslyn managed to get quite a bit of math-comp to work by making Prop a notation for Type.
The same might work for the small fragment of the stdlib we need to get the tactics to work.

**Comment by JasonGross** [comment]
That sounds like a bug.  See if you can construct a minimal example of
failure, and then see if you can get it to fail with the official stdlib?

Though, if your issues are only with -1 and -2, perhaps the issue is that
we have reserved notations for those for trunc indices?  You said you
couldn't get it to work for any negative numbers, but surely you haven't
tried all of them.  Did you try, say, -7?

**Comment by Alizter** [comment]
@JasonGross You are correct about the reservation for trunc indicies! I should learn to count higher. I didn't think this would be an issue since I closed `trunc_scope`. How can I get around the trunc index reservation?

**Comment by Alizter** [comment]
Also I should mention that trying `-10` is now failing, but the error message says: `Unknown interpretation for notation "-1".` So clearly there is a level problem. I can't change the level for `-` since it seems to be reserved at 35. Do you know what level numerals are parsed at? And can I change the reservation of `-`? I am not exactly sure where it is coming from.

**Comment by SkySkimmer** [comment]
>How can I get around the trunc index reservation?

Implement the trunc index notations with numeral notations too, removing the reservation.

**Comment by Alizter** [comment]
@SkySkimmer I am not quite sure how this can be done. Does `Numeral Notation` support negative integers? I still don't understand. I managed to get negative integers working for our int by using `- x` as a notation.

Trunc indices don't have a negation operation I can assign this notation to.

I feel like I might be doing something wrong here.

**Comment by SkySkimmer** [comment]
I don't know what you did, but for instance this works:
~~~coq
Inductive foo := minus_two | plus_four.

Import Decimal.

Definition of_int (x:foo) : Decimal.int :=
  match x with
  | minus_two => (-2)%int (* need to parenthesize otherwise parsed as - (2%int) *)
  | plus_four => 4%int
  end.

Definition to_int (x:Decimal.int) : option foo :=
  match x with
  | (-2)%int => Some minus_two
  | 4%int => Some plus_four
  | _ => None
  end.

Declare Scope foo_scope.
Open Scope foo_scope.

Numeral Notation foo to_int of_int : foo_scope.

Check -2. (* no need for parentheses since we globally opened foo_scope)
Fail Check 5.
~~~

**Comment by Alizter** [comment]
@SkySkimmer I did something funky yesterday but I can't reproduce it. It is all magically working today. I will do as you suggested for trunc index.

I was trying to write out a minimal example with my conversion functions when it worked. Now looking back at what I am working on it is also working. I am very confused but I am happy things are working.

Thanks for your help!

**Comment by Alizter** [comment]
@JasonGross The issue with `-10` seems to also come from reserving `-1` with trunc as I don't have that issue with `-30`.

**Comment by Alizter** [comment]
@SkySkimmer On closer inspection, trunc_index seems to also be a numeral notation, but the following is done straight after:
```coq
(** We declare [-2] and [-1] as notations so that they bind tighter than application, so we can write [IsTrunc -1] rather than [IsTrunc (-1)]. *)
Notation "-2" := (- 2) : trunc_scope.
Notation "-1" := (- 1) : trunc_scope.
```

Do you know how to get -1 and -2 to bind tighter than application, so we don't have brackets?

**Comment by mikeshulman** [comment]
Actually I doubt there's much point to worry about that.  Since `IsHProp` is a notation for `IsTrunc -1` we almost never actually write `IsTrunc -1` (or anything else with a -1 in it) anyway.  Let's just use the brackets.

**Comment by Alizter** [comment]
We still need brackets around when tactics are invoked however. The following fails now:
`exact -2.` we need to instead write `exact (-2).` Not really sure how to fix this, but for now I don't think it is such a big deal.

**Comment by mikeshulman** [comment]
You have to put parentheses around pretty much any other term that gets passed to a tactic, so I don't think putting them around -2 is a problem.

**Comment by Alizter** [comment]
@mikeshulman The tactic one was a bad example. We do write `foo -2` a lot in the library. I have gone ahead and made all these changes. For example in a few places we have written `image -1` etc. I think it is probably safer to keep it as `image (-1)`.

**Comment by Alizter** [comment]
Today's update:

I am very close to finishing the integers. I only have associativity of addition left to show.
 * I have defined the group of integers.
 * I have redefined homotopy groups to use new group theory stuff
 * I have made the HIT circle compatible with the new integers
 * All current calculations of homotopy groups are now group isomorphisms

Once I finish associativity I will clean up and break up into multiple PRs.

**Comment by Alizter** [comment]
Good news! I have finished associativity of the integers. It is quite a non-trivial thing to prove. Not to mention, incredibly boring. The coq proof uses the fact that any two positive integers n, m fall into the following cases: n > m, n = m and n < m. For the inequalities, positive integers can be rewritten as n > m => a + m = n etc.

I really didn't want to go down this route, so I did it just from induction on binary positive integers. One would think that you could prove the "peano induction" for the binary positive integers and show that operations respect succ like in the unary case. However you need associativity of the integers to show this!

My proofs are long and have many cases. They are valid however, and since these are sets I don't think anyone really cares about the content of the proofs. (They are all Qed anyway). Hopefully in the future a cleverer person (or version of myself) can make these slicker than they are.

I will clean up my work and open a PR soon.

**Comment by mikeshulman** [comment]
Is this issue closed by #1137?
